### ISSUE TITLE
[Feature request] Dynamic imports

### ISSUE THREAD

**Issue by streamich** [laverdet/isolated-vm#199]
Thanks for this great library!

Please consider adding dynamic `import()` support.

```js
const something = await import('something');
```

P.S. I would love to create PR for this feature instead of this ticket, but unfortunately my V8 knowledge is not that good.

**Comment by laverdet** [comment]
To update on this I'm currently holding back a little bit on expanding module support because the development within nodejs and v8 is still pretty rapid. I think some semver-major changes will be required to make module support complete. I get the feeling most users are using Webpack to load scripts into isolated-vm as well, so it's kind of niche feature.

**Comment by cyberiums** [comment]
@laverdet Can you give a sample of webpack script for helping beginners understand this?
I have a small program which has to run in another environment but it has certain imports(require).

**Comment by laverdet** [comment]
For example if you wanted lodash available in your isolate you would put this in `webpack.config.js`:

```js
module.exports = {
  entry: 'lodash',
  output: {
    library: 'lodash',
  },
};
```

After running `npx webpack` you'll find `dist/main.js` which is a packaged lodash file. If you do `isolate.compileScriptSync(fs.readFileSync('./dist/main.js', 'utf8')).runSync(context)` now there will be a global variable `lodash` in the context which is what the return result of `require('lodash')` would have been.

That should be enough to get you started.

**Comment by theoutlander** [comment]
I don't think this works when using the typescript compiler directly. However, supporting dynamic imports will open up security concerns.

**Comment by laverdet** [comment]
Security concerns?

**Comment by roobscoob** [comment]
Any updates here? Working on an Open Source project for game modding that uses this library and the lack of Dynamic Imports breaks a lot for us :/

**Comment by idleman** [comment]
If you really need dynamic import must it be implemented in the backend (C++ code). However, is a basic "require" function good enough can you do the following:

1. Create a global __require in a selected Isolate. The function should take a string and return the source code of a specific file. You can use the Reference and Context API to create this function. See examples.
2. Create a wrapper function: "require" that calls __require() and then evaluate the code. I would use "new Function()". This way you should be able to control the "this" property. Something like:

```

function require(specifier) {
  const code = __require(specifier); // you need to implement this function using the isolated-vm API. It should return the source code
  const construct = new Function(`
    const module = this;
    const exports = {};
    module.exports = exports;
    ${code}
    return module;
  `)
  return construct.call({});
}
```

**Comment by jakedaleweb** [comment]
Just to build on @idleman 's excellent comment, I've recently been successful in importing a module (https://www.npmjs.com/package/buffer) to the isolate with this method, here's what I did - would love a more official way to do this though.

```
    // Specify every module that will be 'require'd including chaining dependencies
    // i.e. buffer requires base64-js
    const requiredModules = ['buffer/', 'base64-js', 'ieee754']

    // Store the module code in a string
    const moduleStrings = requiredModules.reduce((mut_acc, curr) => {
      mut_acc[curr] = fs.readFileSync(`./node_modules/${curr}/index.js`, 'utf8')
      return mut_acc
    }, {} as Record<string, string>)

    // Make the module code strings copyable to the sandbox and copy them in
    const externalCopy = new ivm.ExternalCopy(moduleStrings).copyInto({release: true})
    sandbox.setSync('moduleStrings', externalCopy)

    // Add and call a special require function
    // call it for buffer and set 'Buffer' const
    context.evalSync(`
      function require(specifier) {
        const code = moduleStrings[specifier]
        const construct = new Function(\`
          const module = this;
          const exports = {};
          module.exports = exports;
          \${code}
          return module.exports;
        \`)
        return construct.call({})
      }
      const Buffer = require('buffer/').Buffer
    `)
```

**Comment by daniel-santos** [comment]
I have dynamic module loading working to a degree and with a few caveats.

1. I didn't design the code as a generic solution, and the security implementation is basic.
2. it expects the modules and all of their dependencies to exist in the file system at `$(pwd)/node_modules`, but a robust implementation should allow the caller to specify exactly how modules are loaded and supply security information. In my case, the code just doesn't allow you to escape the `node_modules` sub-directory, but a proper implementation will be more restrictive.

In my case, I build an `allowedModules` map of module names to the parsed package.json as an argument to my `IVMLoader`'s constructor, along with the sources and a file name for the script to be run. But since I'm already mucking with the file system, I could have just as well just pass the module names and let the `IVMLoader` class fetch the package.json, since that's what it does for dependencies anyway.

I throw a distinct `ScriptError` for compilation failures and errors loading dependencies (like trying to import a symbol from a module that isn't exported), with any other error being from the script's execution it's self. My use case requires concealing details of the framework (á la "security through obscurity"), so I catch `ScriptError` and only return the `message`, but for other errors I scrub the stack trace and return the `message` along with the IVM portion of the stack trace.

In the `IVMLoader` class, I manually error on circular dependencies and cache everything so it never tries to load a module twice (like if it's imported in more than one place).

To make this a full-fledged solution, I think it would need to be passed a file system, compressed archive or such that contains only the modules and code that are allowed to be loaded.

**Comment by d3x0r** [comment]
where would a handler for `import()` go ?  There's a function hooked in somewhere that JS code ends up calling that returns a promise, and throws an Error with .message= `Not supported Error: Not supported` and I can't find where that error is even generated...  static imports work in the same code string, so it's definitely running a module... 

`import( 'specifier' ).then( ... ) .catch( err=> console.log( err.message ) );`  -> `Not supported Error: Not supported` 
but `import blah from "specifier"` in the same file does work.
